src/MSRFieldControl.cpp

```diff
@@ -0,0 +1,146 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+
+#include "MSRFieldControl.hpp"
+#include "MSRIO.hpp"
+#include "MSR.hpp"
+#include "Exception.hpp"
+#include "geopm_debug.hpp"
+
+namespace geopm
+{
+    MSRFieldControl::MSRFieldControl(std::shared_ptr<MSRIO> msrio,
+                                     int cpu,
+                                     uint64_t offset,
+                                     int begin_bit,
+                                     int end_bit,
+                                     int function,
+                                     double scalar)
+        : m_msrio(msrio)
+        , m_cpu(cpu)
+        , m_offset(offset)
+        , m_shift(begin_bit)
+        , m_num_bit(end_bit - begin_bit + 1)
+        , m_mask(((1ULL << m_num_bit) - 1) << begin_bit)
+        , m_function(function)
+        , m_scalar(scalar)
+        , m_inverse(1.0 / scalar)
+        , m_is_batch_ready(false)
+        , m_adjust_idx(-1)
+    {
+        if (m_msrio == nullptr) {
+            throw Exception("MSRFieldControl: cannot construct with null MSRIO",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (m_function < 0 || m_function >= MSR::M_FUNCTION_OVERFLOW) {
+            throw Exception("MSRFieldControl: unsupported encode function.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (begin_bit > end_bit) {
+            throw Exception("MSRFieldControl: begin bit must be <= end bit",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+    }
+
+    void MSRFieldControl::setup_batch(void)
+    {
+        GEOPM_DEBUG_ASSERT(m_msrio != nullptr, "null MSRIO");
+        if (!m_is_batch_ready) {
+            m_adjust_idx = m_msrio->add_write(m_cpu, m_offset);
+            m_is_batch_ready = true;
+        }
+    }
+
+    uint64_t MSRFieldControl::encode(double value) const
+    {
+        uint64_t result = 0;
+        double value_inferred = 0.0;
+        uint64_t float_y, float_z;
+        switch (m_function) {
+            case MSR::M_FUNCTION_SCALE:
+                result = (uint64_t)(m_inverse * value);
+                break;
+            case MSR::M_FUNCTION_LOG_HALF:
+                // F = S * 2.0 ^ -X =>
+                // X = log2(S / F)
+                result = (uint64_t)std::log2(m_scalar / value);
```

Comment:
result = (uint64_t)(-1.0*std::log2(m_inverse * value));

Reply:
will address this in the next patch adding save/control support